File: typescript/src/projects/apiKeys/resource.ts
import { Api, Endpoints, HttpMethods } from "../../api";
import { ApiKey, DeletionResult } from "../../types";
import {
  ApiKeyHttpResponse,
  ApiKeysHttpResponse,
  CreateApiKeyRequest,
  UpdateApiKeyRequest,
} from "./types";

export class ProjectApiKeys {
  constructor(
    private readonly api: Api,
    private readonly projectId: string,
  ) {}

  async list(): Promise<ApiKey[]> {
    const data = await this.api.sendRequest<ApiKeysHttpResponse>(
      HttpMethods.GET,
      Endpoints.PROJECT_API_KEYS_ENDPOINT,
      { urlParams: { projectId: this.projectId } },
    );
    return data.apiKeys ?? [];
  }

  async get(apiKeyId: number): Promise<ApiKey> {
    const data = await this.api.sendRequest<ApiKeyHttpResponse>(
      HttpMethods.GET,
      Endpoints.PROJECT_API_KEY_ENDPOINT,
      { urlParams: { projectId: this.projectId, apiKeyId } },
    );
    return data.apiKey;
  }

  async create(params: CreateApiKeyRequest): Promise<ApiKey> {
    const body: CreateApiKeyRequest = { name: params.name };
    const data = await this.api.sendRequest<ApiKeyHttpResponse>(
      HttpMethods.POST,
      Endpoints.PROJECT_API_KEYS_ENDPOINT,
      { body, urlParams: { projectId: this.projectId } },
    );
    return data.apiKey;
  }

  async update(
    apiKeyId: number,
    params: UpdateApiKeyRequest,
  ): Promise<ApiKey> {
    const body: UpdateApiKeyRequest = { valid: params.valid };
    const data = await this.api.sendRequest<ApiKeyHttpResponse>(
      HttpMethods.PUT,
      Endpoints.PROJECT_API_KEY_ENDPOINT,
      { body, urlParams: { projectId: this.projectId, apiKeyId } },
    );
    return data.apiKey;
  }

  async delete(apiKeyId: number): Promise<DeletionResult> {
    return this.api.sendRequest<DeletionResult>(
      HttpMethods.DELETE,
      Endpoints.PROJECT_API_KEY_ENDPOINT,
      { urlParams: { projectId: this.projectId, apiKeyId } },
    );
  }
}

File: python/confidentai/projects/api_keys/resource.py
from typing import List

from ...api import Api, Endpoints, HttpMethods
from ...types import ApiKey, DeletionResult
from .types import (
    ApiKeyHttpResponse,
    ApiKeysHttpResponse,
    CreateApiKeyRequest,
    UpdateApiKeyRequest,
)


class ProjectApiKeys:
    def __init__(self, api: Api, project_id: str) -> None:
        self._api = api
        self._project_id = project_id

    def list(self) -> List[ApiKey]:
        data, _ = self._api.send_request(
            HttpMethods.GET,
            Endpoints.PROJECT_API_KEYS_ENDPOINT,
            url_params={"projectId": self._project_id},
        )
        return ApiKeysHttpResponse(**data).api_keys

    def get(self, api_key_id: int) -> ApiKey:
        data, _ = self._api.send_request(
            HttpMethods.GET,
            Endpoints.PROJECT_API_KEY_ENDPOINT,
            url_params={"projectId": self._project_id, "apiKeyId": api_key_id},
        )
        return ApiKeyHttpResponse(**data).api_key

    def create(self, name: str) -> ApiKey:
        body = CreateApiKeyRequest(name=name).model_dump(
            by_alias=True, exclude_none=True
        )
        data, _ = self._api.send_request(
            HttpMethods.POST,
            Endpoints.PROJECT_API_KEYS_ENDPOINT,
            body=body,
            url_params={"projectId": self._project_id},
        )
        return ApiKeyHttpResponse(**data).api_key

    def update(self, api_key_id: int, *, valid: bool) -> ApiKey:
        body = UpdateApiKeyRequest(valid=valid).model_dump(
            by_alias=True, exclude_none=True
        )
        data, _ = self._api.send_request(
            HttpMethods.PUT,
            Endpoints.PROJECT_API_KEY_ENDPOINT,
            body=body,
            url_params={"projectId": self._project_id, "apiKeyId": api_key_id},
        )
        return ApiKeyHttpResponse(**data).api_key

    def delete(self, api_key_id: int) -> DeletionResult:
        data, _ = self._api.send_request(
            HttpMethods.DELETE,
            Endpoints.PROJECT_API_KEY_ENDPOINT,
            url_params={"projectId": self._project_id, "apiKeyId": api_key_id},
        )
        return DeletionResult(**data)


class AsyncProjectApiKeys:
    def __init__(self, api: Api, project_id: str) -> None:
        self._api = api
        self._project_id = project_id

    async def list(self) -> List[ApiKey]:
        data, _ = await self._api.a_send_request(
            HttpMethods.GET,
            Endpoints.PROJECT_API_KEYS_ENDPOINT,
            url_params={"projectId": self._project_id},
        )
        return ApiKeysHttpResponse(**data).api_keys

    async def get(self, api_key_id: int) -> ApiKey:
        data, _ = await self._api.a_send_request(
            HttpMethods.GET,
            Endpoints.PROJECT_API_KEY_ENDPOINT,
            url_params={"projectId": self._project_id, "apiKeyId": api_key_id},
        )
        return ApiKeyHttpResponse(**data).api_key

    async def create(self, name: str) -> ApiKey:
        body = CreateApiKeyRequest(name=name).model_dump(
            by_alias=True, exclude_none=True
        )
        data, _ = await self._api.a_send_request(
            HttpMethods.POST,
            Endpoints.PROJECT_API_KEYS_ENDPOINT,
            body=body,
            url_params={"projectId": self._project_id},
        )
        return ApiKeyHttpResponse(**data).api_key

    async def update(self, api_key_id: int, *, valid: bool) -> ApiKey:
        body = UpdateApiKeyRequest(valid=valid).model_dump(
            by_alias=True, exclude_none=True
        )
        data, _ = await self._api.a_send_request(
            HttpMethods.PUT,
            Endpoints.PROJECT_API_KEY_ENDPOINT,
            body=body,
            url_params={"projectId": self._project_id, "apiKeyId": api_key_id},
        )
        return ApiKeyHttpResponse(**data).api_key

    async def delete(self, api_key_id: int) -> DeletionResult:
        data, _ = await self._api.a_send_request(
            HttpMethods.DELETE,
            Endpoints.PROJECT_API_KEY_ENDPOINT,
            url_params={"projectId": self._project_id, "apiKeyId": api_key_id},
        )
        return DeletionResult(**data)
